src/main/java/com/mongodb/jdbc/MongoResultSetMetaData.java

```diff
@@ -5,48 +5,122 @@
     protected final int unknownLength = 0;
 
-    public MongoResultSetMetaData(MongoResultDoc metadataDoc) {
-        columns = metadataDoc.columns;
-
-        columnPositions = new HashMap<>(columns.size());
-        int i = 0;
-        for (Column c : columns) {
-            columnPositions.put(c.columnAlias, i++);
-        }
-    }
-
     protected void checkBounds(int i) throws SQLException {
         if (i > getColumnCount()) {
             throw new SQLException("Index out of bounds: '" + i + "'.");
         }
     }
 
-    public int getColumnPositionFromLabel(String label) {
-        return columnPositions.get(label);
+    @Override
+    public boolean isAutoIncrement(int column) throws SQLException {
+        checkBounds(column);
+        return false;
     }
 
-    public boolean hasColumnWithLabel(String label) {
-        return columnPositions.containsKey(label);
+    @Override
+    public boolean isSearchable(int column) throws SQLException {
+        checkBounds(column);
+        return true;
     }
 
     @Override
-    public boolean isAutoIncrement(int column) throws SQLException {
+    public boolean isCurrency(int column) throws SQLException {
+        checkBounds(column);
+        return false;
+    }
+
+    @Override
+    public String getSchemaName(int column) throws SQLException {
+        checkBounds(column);
+        return "";
+    }
+
+    public abstract MongoColumnInfo getColumnInfo(int column) throws SQLException;
+
+    public abstract boolean hasColumnWithLabel(String label) throws SQLException;
+
+    public abstract int getColumnPositionFromLabel(String label) throws SQLException;
+
+    @Override
+    public boolean isReadOnly(int column) throws SQLException {
+        checkBounds(column);
+        return true;
+    }
+
+    @Override
+    public boolean isWritable(int column) throws SQLException {
         checkBounds(column);
         return false;
     }
 
+    @Override
+    public boolean isDefinitelyWritable(int column) throws SQLException {
+        checkBounds(column);
+        return false;
+    }
+
+    @Override
+    public int isNullable(int column) throws SQLException {
+        return getColumnInfo(column).getNullability();
+    }
+
+    @Override
+    public String getColumnLabel(int column) throws SQLException {
+        return getColumnInfo(column).getColumnAlias();
+    }
+
+    @Override
+    public String getColumnName(int column) throws SQLException {
+        return getColumnInfo(column).getColumnName();
+    }
+
+    @Override
+    public String getTableName(int column) throws SQLException {
+        return getColumnInfo(column).getTableAlias();
+    }
+
+    @Override
+    public String getCatalogName(int column) throws SQLException {
+        return getColumnInfo(column).getDatabase();
+    }
+
+    @Override
+    public int getColumnType(int column) throws SQLException {
+        MongoColumnInfo ci = getColumnInfo(column);
+        return ci.getJDBCType();
+    }
```

Comment:
[nit] Why are some methods implemented in one line, like `getCatalogName`, and some broken up into two lines, like `getColumnType`? Can we change them all to be one way or the other?

Reply:
Various stages of refactor ;) Sure.